darts/utils/data/encoder_base.py

```diff
@@ -0,0 +1,292 @@
+        """
+        Implement a method that extracts the required covariate index for prediction.
+
+        Parameters
+        ----------
+        n
+            the forecast horizon
+        target
+            the target TimeSeries used during training or passed to prediction as `series`
+        covariate
+            optionally, the future covariates used for prediction
+        """
+        pass
+
+
+class PastCovariateIndexGenerator(CovariateIndexGenerator):
+    """generates index for past covariate"""
+    def generate_train_series(self,
+                              target: TimeSeries,
+                              covariate: Optional[TimeSeries] = None) -> SupportedIndexes:
+        # TODO -> we can probably summarize this in CovariateIndexGenerator as I think it's the same for
+        #  future & past covs
+        super(PastCovariateIndexGenerator, self).generate_train_series(target, covariate)
+        return covariate.time_index if covariate is not None else target.time_index
+
+    def generate_inference_series(self,
+                                  n: int,
+                                  target: TimeSeries,
+                                  covariate: Optional[TimeSeries] = None) -> SupportedIndexes:
+        """For prediction (`n` is given) with past covariates we have to distinguish between two cases:
+        1)  if past covariates are given, we can use them as reference
+        2)  if past covariates are missing, we need to generate a time index that starts `input_chunk_length`
+            before the end of `target` and ends `max(0, n - output_chunk_length)` after the end of `target`
+        """
+
+        super(PastCovariateIndexGenerator, self).generate_inference_series(n, target, covariate)
+        if covariate is not None:
+            return covariate.time_index
+        else:
+            return _generate_index(start=target.end_time() - target.freq * (self.input_chunk_length - 1),
+                                   length=self.input_chunk_length + max(0, n - self.output_chunk_length),
+                                   freq=target.freq)
+
+
+class FutureCovariateIndexGenerator(CovariateIndexGenerator):
+    """generates index for future covariate."""
+    def generate_train_series(self,
+                              target: TimeSeries,
+                              covariate: Optional[TimeSeries] = None) -> SupportedIndexes:
+        """For training (when `n` is `None`) we can simply use the future covariates (if available) or target as
+        reference to extract the time index.
+        """
+
+        # TODO -> we can probably summarize this in CovariateIndexGenerator as I think it's the same for
+        #  future & past covs
+        super(FutureCovariateIndexGenerator, self).generate_train_series(target, covariate)
+
+        return covariate.time_index if covariate is not None else target.time_index
```

Comment:
I think this is correct for training.

For the training dataset, the last `output_chunk_length` points of the target series are never used as a "past/input" chunk because we need to evaluate the loss on the future chunk.
So the future covariates can end at the same index as the target.

But yeah maybe I missed something, WDYT?

Reply:
Maybe it would be smarter though to include the next `output_chunk_length` points. Then we wouldn't have to call `encode_inference()` in the TrainingDatasets. Then we just have `output_chunk_length` unused future covariates when adding future covariates in the "eager/non-lazy" way.

Reply:
since we don't distinguish between lazy and eager case anymore, this can stay as it is